### ISSUE TITLE
Feature request: allow the user to pass a precompile script to the build image task

### ISSUE THREAD

**Issue by ericphanson** [julia-vscode/julia-vscode#1160]
The function https://julialang.github.io/PackageCompiler.jl/dev/refs/#PackageCompiler.create_sysimage takes a keyword argument `precompile_execution_file` which can run some commands to precompile them. It would be nice if creating a `VSCodePrecompileExecutionFile.jl` in the working directory were enough (or somehow through the run task UI). That way you could eg add Makie to your project and do some plotting in that file, and hopefully have a fast TTFP that way.

**Comment by davidanthoff** [comment]
Wouldn't it make most sense that there is some standard location and filename for such files in packages, and that `PackageCompiler.create_sysimage` automatically looks for such a file in the packages one aims to compile, and then automatically uses them?

I would much prefer to not add configuration options around this on the VS Code side of things.

I also think it would not be ideal to look in the current workspace for such a file: the sysimage compile command really takes as an input an environment, which is pretty unrelated to whatever is happening in the current workspace.

I should say that I'm very much in favor of the goal!

**Comment by ericphanson** [comment]
> Wouldn't it make most sense that there is some standard location and filename for such files in packages, and that PackageCompiler.create_sysimage automatically looks for such a file in the packages one aims to compile, and then automatically uses them?

I think this makes sense also, but is a bit orthogonal, because such files would only be relevant for using functions within a single package whereas a file in the workspace could contain commands that use multiple packages together (e.g. a function from one package with a type from another).

To me it makes sense to have this in the workspace to have a sort of story about "compile a sysimage for the project I'm working on + precompile some commands from dependencies specific to this project". In other words, to me one of the cool features of integrating package compilation into the IDE is having very specific sysimages customized to the particular set of packages I'm using for a certain project, and being able to precompile commands using that set of packages in a way that's particular for the project makes sense.

E.g. if I were solving differential equations with some slightly unusual numeric type like DoubleF64s it wouldn't make sense for either package to include that combination in their own `precompile_execution_file`s (if a standard develops around putting those in packages), but I could solve a few in my project-specific `precompile_execution_file` which could save a lot of compile time. (Maybe more relevant for me personally would be a high-precision type + an optimization problem, but it's the same thing).

> I would much prefer to not add configuration options around this on the VS Code side of things.

That's fair-- if a standard spot does develop (and I know e.g. @DilumAluthge is interested in that), then the same could be used here in the VS code project. (So e.g. each dependency package could have such a file, and the application/user project could also have one-- the latter being the one I'm talking about in this issue). I think the VS Code task would still need to pass the location along to PackageCompiler.jl but it wouldn't need to be a config option if there's a standard location to check.

**Comment by davidanthoff** [comment]
So maybe we should say that if a `precompile_execution_file.jl` file exists parallel to the `Project.toml` and `Manifest.toml` of the env that one is precompiling a sysimage for, it will be used? That way we would keep the idea that a sysimage is for a specific env? And if such a file exists in a package that is compiled into the sysimage, it is also used?

I would still love if these kinds of design decisions were made in PackageCompiler.jl, and we could just call a function `PackageCompiler.create_sysimg_for_env(path_to_env)` and then it would pick up these files etc., without us having to make VS Code specific filename/location designs.

**Comment by ericphanson** [comment]
> So maybe we should say that if a precompile_execution_file.jl file exists parallel to the Project.toml and Manifest.toml of the env that one is precompiling a sysimage for, it will be used? That way we would keep the idea that a sysimage is for a specific env? And if such a file exists in a package that is compiled into the sysimage, it is also used?

Yeah, this is exactly what I was thinking.

> I would still love if these kinds of design decisions were made in PackageCompiler.jl, and we could just call a function PackageCompiler.create_sysimg_for_env(path_to_env) and then it would pick up these files etc., without us having to make VS Code specific filename/location designs.

Ok, that makes sense. I will open an issue there to discuss automatically picking up the file from a standard location to see if we can do this that way. So you think it's best if the PackageCompiler API supports passing a directory, and then PackageCompiler itself uses the included Project + Manifest to get the packages to compile and looks for a `precompile_execution_file.jl`?

**Comment by davidanthoff** [comment]
Yeah, so in my ideal world I could change [this line](https://github.com/julia-vscode/julia-vscode/blob/f76fa7bf2e17dc4b6ce8fd871891a8bf65c8d6c0/scripts/tasks/task_compileenv.jl#L15) to:
```julia
PackageCompiler.create_sysimage(sysimage_path=sysimage_path, project=env_to_precompile)
```
and then PackageCompiler.jl would just do whatever is the most sensible thing, i.e. precompile all the packages in that project into the sysimage, run whatever scripts it might want to run that are in standard locations etc.
